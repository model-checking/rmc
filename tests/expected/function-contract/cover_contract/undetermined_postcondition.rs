// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

// Test `cover_contract` functionality, which fails verification for a unsatisfiable precondition
// or unreachable postcondition.
// See https://github.com/model-checking/kani/issues/2793

// Undetermined whether we can reach the postcondition because we encounter an unsupported construct.

#[kani::proof_for_contract(unsupp)]
fn undetermined_example() {
    let mut x = 0;
    unsupp(&mut x);
    assert!(x == 0);
}

#[kani::requires(true)]
#[kani::ensures(|res| *res == *x)]
fn unsupp(x: &mut u8) -> u8 {
    unsafe {
        std::arch::asm!("nop");
    }
    *x
}
