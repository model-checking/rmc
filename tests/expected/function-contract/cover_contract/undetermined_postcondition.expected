
contract_cover.2\
	 - Status: UNDETERMINED\
	 - Description: "The contract's postcondition is reachable."

unsupported_construct.1\
	 - Status: FAILURE

VERIFICATION:- FAILED